nemo/collections/nlp/modules/common/megatron/retrieval_service.py

```diff
@@ -97,3 +207,272 @@ def get_knn(self, query: Union[List[str], str, torch.Tensor]):
+        self.no_retrieval = np.ones((1, 1, 2 * self.chunk_size), dtype=ds._index.dtype) * pad_id
+        if torch.distributed.get_rank() == 0:
+            server = RetrievalServer(faiss_index, faiss_devices, nprobe, retrieval_index, tokenizer)
+            server.run("0.0.0.0")
+        torch.distributed.barrier()
+
+    def get_knn(self, query: Union[List[str], str, torch.Tensor], neighbors):
+        if isinstance(query, torch.Tensor):
+            sentence_list = []
+            for q in query:
+                text = self.tokenizer.ids_to_text(q)
+                sentence_list.append(text)
+            query = sentence_list
+        if neighbors == 0:
+            # use padding
+            return np.repeat(self.no_retrieval, len(query), 0).astype(np.int64)
+        data = {'sentences': query}
+        data['neighbors'] = neighbors
+        result = request_data(data, PORT_NUM)
+        result = np.array(result)
+        return result
+
+
+class DynamicFaissRetrievalService(RetrievalService):
+    def __init__(
+        self, faiss_devices: str, tokenizer: TokenizerSpec, chunk_size: int, stride: int,
+    ):
+        self.updatable = True
+        self.tokenizer = tokenizer
+        self.chunk_size = chunk_size
+        pad_id = self.tokenizer.pad_id
+        # batch, neighbors, 2*chunk_size
+        self.no_retrieval = np.ones((1, 1, 2 * self.chunk_size), dtype=np.int64) * pad_id
+        if torch.distributed.get_rank() == 0:
+            server = DynamicRetrievalServer(faiss_devices, tokenizer, chunk_size, stride)
+            server.run("0.0.0.0")
+        torch.distributed.barrier()
+
+    def get_knn(self, query: Union[List[str], str, torch.Tensor], neighbors):
+        if isinstance(query, torch.Tensor):
+            sentence_list = []
+            for q in query:
+                text = self.tokenizer.ids_to_text(q)
+                sentence_list.append(text)
+            query = sentence_list
+        if neighbors == 0:
+            # use padding
+            return np.repeat(self.no_retrieval, len(query), 0).astype(np.int64)
+        data = {'sentences': query}
+        data['neighbors'] = neighbors
+        result = request_data(data, PORT_NUM_DYN)
+        result = np.array(result)
+        return result
+
+    def add_docs_to_index(self, query: List[str], add_eos: bool = True):
+        """
+        Add documents to the Faiss index
+        Args:
+            docs: List[str], list of documents that is going to be added to the index
+            add_eos: bool, whether add the eos in the end
+        """
+        if isinstance(query, torch.Tensor):
+            sentence_list = []
+            for q in query:
+                text = self.tokenizer.ids_to_text(q)
+                sentence_list.append(text)
+            query = sentence_list
+        data = {'sentences': query, 'add_eos': add_eos}
+        return request_data(data, PORT_NUM_DYN)
+
+
+class ComboRetrievalService(RetrievalService):
```

Comment:
Another comment here would be helpful. In general the differences of use cases.

Reply:
added comments.